python/fledge/services/core/api/service.py

```diff
@@ -183,29 +185,72 @@ async def add_service(request):
                 storage = connect.get_storage_async()
-                audit = AuditLogger(storage)
-                audit_detail = {'packageName': name}
-                log_code = 'PKGUP' if msg == 'updated' else 'PKGIN'
-                await audit.information(log_code, audit_detail)
-                return web.json_response({'message': message, "link": link})
+                action = "install"
+                select_payload = PayloadBuilder().SELECT("status").WHERE(['action', '=', action]).AND_WHERE(
+                    ['name', '=', name]).payload()
+                result = await storage.query_tbl_with_payload('packages', select_payload)
+                response = result['rows']
+                if response:
+                    exit_code = response[0]['status']
+                    if exit_code == -1:
+                        msg = "{} package installation already in progress".format(name)
+                        return web.HTTPTooManyRequests(reason=msg, body=json.dumps({"message": msg}))
+                    # Remove old entry from table for other cases
+                    delete_payload = PayloadBuilder().WHERE(['action', '=', action]).AND_WHERE(
+                        ['name', '=', name]).payload()
+                    await storage.delete_from_tbl("packages", delete_payload)
+
+                # Check If requested service is already installed and then return immediately
+                services = get_service_installed()
+                svc_name = name.split('fledge-')[1].split('-')[1]
+                for s in services:
+                    if s == svc_name:
+                        msg = "{} package is already installed".format(name)
+                        return web.HTTPNotModified(reason=msg, body=json.dumps({"message": msg}))
+
+                # Check If requested service is available for configured repository
+                services, log_path = await common.fetch_available_packages("service")
+                if name not in services:
+                    raise KeyError('{} service is not available for the given repository'.format(name))
+                # Insert record into Packages table
+                insert_payload = PayloadBuilder().INSERT(id=str(uuid.uuid4()), name=name, action=action, status=-1,
+                                                         log_file_uri="").payload()
+                result = await storage.insert_into_tbl("packages", insert_payload)
+                response = result['response']
+                if response:
+                    # GET id from Packages table to track the installation response
+                    select_payload = PayloadBuilder().SELECT("id").WHERE(['action', '=', action]).AND_WHERE(
+                        ['name', '=', name]).payload()
+                    result = await storage.query_tbl_with_payload('packages', select_payload)
+                    response = result['rows']
```

Comment:
why select after insert, don't we already know the inserted items?

Reply:
INSERT returns only affected rows from storage. And SELECT needs to get id which we have used further...

Reply:
On success of INSERT, instead of select we can assign/use same id `.INSERT(id=str(uuid.uuid4()),` instead of SELECT?